### PR TITLE
Add standings calculations

### LINKED ISSUE #24: Add standings calculation
We want to calculate wins and losses and total score of a team given the data in the scores table

### PR DESCRIPTION
**Pull request by adamszcz**
# What I added (link any issues addressed)
A controller and service that will handle the processing of standings based on the stored data

# How to test it
- Call ```/standings/calculate``` with a Team object in the body
- See that a Team object is returned
- Ensure that there is at least one Game in the database with a Winner recorded
- Run the endpoint against one of those teams, and the standings should reflect the changes

### CODE REVIEW

File: CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            Game game = new Game(){ Teams = new List<Team>(){team1, team2}};
+            var games = new List<Game>(){game};
+            mockGameRepository.Setup(x => x.GetByTeamId(team1.Id)).ReturnsAsync(games);
+            var result = await StandingsService.Calculate(team1);
+            Assert.IsType<List<Game>>(result.BracketGames);
+            Assert.IsType<List<Game>>(result.PlayInGames);
+        }
+        [Fact]
+        public async Task Calculate_ShouldReturnTeamWithOneWinIfTheyHaveLoss()
```
Comment (cszczepaniak) [inline CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs]: 🤔 I'm a bit confused by this test name. Which team is supposed to have one win? Why are we not asserting wins?

File: src/CribblyBackend/Controllers/StandingsController.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+using Serilog;
+
+namespace CribblyBackend.Controllers
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    public class StandingsController : ControllerBase
+    {
+        private readonly IStandingsService StandingsService;
+        private readonly ILogger logger;
```
Comment (cszczepaniak) [inline src/CribblyBackend/Controllers/StandingsController.cs]: 🌵🌵 We've been a bit inconsistent with our naming schemes. Which one we go with is a matter of preference, but whatever we choose we should remain consistent. FWIW my preference is to have private fields be `_underscoreCamelCase` - the advantage with this is that the constructor doesn't need to use `this.` to differentiate between fields and parameters.

Either way though, let's pick one and stick with it 🙂

File: src/CribblyBackend/Services/StandingsService.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            _gameRepository = gameRepository;
+        }
+        public async Task<Team> Calculate(Team team)
+        {
+            var allGames = await _gameRepository.GetByTeamId(team.Id);
+            team.PlayInGames = allGames.Where(g => g.GameRound <= Game.Round.Round3).ToList();
+            team.BracketGames = allGames.Where(g => g.GameRound >= Game.Round.TourneyRound1).ToList();
+
+            team.Wins = allGames.Where(g => g.Winner != null && g.Winner.Name == team.Name).ToList().Count;
+            team.Losses = allGames.Where(g => g.Winner != null && g.Winner.Name != team.Name).ToList().Count;
```
Comment (cszczepaniak) [inline src/CribblyBackend/Services/StandingsService.cs]: I'm not _exactly_ sure on this, but I think the performance might be slightly better if you omit `ToList` and simply call `allGames.Count(g => g.Winner != null && g.Winner.Name == team.Name);`

Also, might it be better to use `team.Id` and `g.Winner.Id` since `Id` is guaranteed to be unique?

File: src/CribblyBackend/Services/TeamService.cs [hunk 4]
```diff
@@ -42,7 +39,8 @@ public void Delete(Team team)
 
         public async Task<Team> GetById(int id)
         {
-            return await _teamRepository.GetById(id);
+            var team = await _teamRepository.GetById(id);
+            return team;
```
Comment (cszczepaniak) [inline src/CribblyBackend/Services/TeamService.cs]: Why this change?

**Review by cszczepaniak — APPROVED** [review]
A few small comments/suggestions, but I won't hold this one up :+1:

**Comment by cszczepaniak** [comment]
Oh... also somehow the build is failing?

**Comment by adamszcz** [comment]
Yeah I am not sure why this build is failing. Trying to figure that out. The logs reference something that doesn't even exist anymore

**Comment by cszczepaniak** [comment]
Yup, looks like `GameService.cs` line 53..... which as you say doesn't exist 🙃

File: CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs (hunk 1 above)
Comment (adamszcz) [reply CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs]: This is basically asserting that a team that has lost their only game comes back as 0-1. Maybe I should change the test name?

File: src/CribblyBackend/Services/StandingsService.cs (hunk 3 above)
Comment (adamszcz) [reply src/CribblyBackend/Services/StandingsService.cs]: Agreed and fixed!

File: src/CribblyBackend/Controllers/StandingsController.cs (hunk 2 above)
Comment (adamszcz) [reply src/CribblyBackend/Controllers/StandingsController.cs]: Fixed and agreed!

File: src/CribblyBackend/Services/TeamService.cs (hunk 4 above)
Comment (adamszcz) [reply src/CribblyBackend/Services/TeamService.cs]: Not sure why actually, lol. I switched it back

**Review by adamszcz — COMMENTED** [review]
Wait, mine is showing ```TeamService.cs``` on line 52...

File: CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs (hunk 1 above)
Comment (cszczepaniak) [reply CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs]: As far as I can see you're asserting that the team has one loss, so I'd expect the test to be named something more like `Calculate_ShouldReturnTeamWIthOneLossIfTheyHaveLoss()`

File: src/CribblyBackend/Services/StandingsService.cs (hunk 3 above)
Comment (adamszcz) [reply src/CribblyBackend/Services/StandingsService.cs]: So the tests keep failing when I use ```team.Id``` and ```g.Winner.Id```. Will need to keep tinkering to get that to work

Comment (cszczepaniak) [reply src/CribblyBackend/Services/StandingsService.cs]: Probably because your test teams both have the same Id 🙂 
https://github.com/cszczepaniak/cribbly-backend/blob/6fd1e55857b2ad6543b3c61ea901ec4cc3adf5bc/CribblyBackend.UnitTests/ServiceTests/StandingsServiceTests.cs#L17-L18

Comment (adamszcz) [reply src/CribblyBackend/Services/StandingsService.cs]: Nvm, figured it out - wrote the test wrong!
